Source/WebCore/rendering/RenderLayer.h

```diff
@@ -1180,6 +1193,11 @@ class RenderLayer : public CanMakeWeakPtr<RenderLayer> {
     // Convert a point in absolute coords into layer coords, taking transforms into account
     LayoutPoint absoluteToContents(const LayoutPoint&) const;
 
+    bool hasNonOpacityTransparency() const
+    {
+        return renderer().hasMask() || hasBlendMode() || (isolatesBlending() && !renderer().isDocumentElementRenderer());
```

Comment:
Could a filter also modify opacity? Is this a point at which a filter would be applied?

Reply:
A filter can also apply opacity, but filters are handled separately, so this change won't affect them.

There's definitely the potential to do a similar optimisation with filters (skip creating a temporary group when the filter could be applied without it).